### PR TITLE
Add single-series counter events

### PR DESCRIPTION
**Pull request by efernandez**
See https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#heading=h.msg3086636uq

Counter name is built as `<counter_name> + '_' + <series_name>`

### CODE REVIEW

File: scalopus_tracing/include/scalopus_tracing/internal/counter_tracepoint.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace scalopus
+{
+/**
+ * @brief Emits a counter event.
+ * @param id The tracepoint id to relate it to a human readable string.
+ * @param counter_name The counter name.
+ * @param counter_value The counter value.
+ */
+void counter_event(const unsigned int id, const std::string& counter_name, const unsigned int counter_value);
```
Comment (iwanders) [inline scalopus_tracing/include/scalopus_tracing/internal/counter_tracepoint.h]: At first I was going to comment that `counter_name` is superfluous as it should be stored in the `id` via the [StaticStringTracker](https://github.com/iwanders/scalopus/blob/master/scalopus_tracing/include/scalopus_tracing/internal/static_string_tracker.h#L45).

However, then I started experimenting a bit with a trace file. Basically you can have a counter with a name, and in that single counter you can have multiple series. 

See: [trace_with_counters.json.gz](https://github.com/iwanders/scalopus/files/3643607/trace_with_counters.json.gz)
Where I made three counter rows, one with animal counts, one with animal sounds. Both have entries for cats and dogs. I also tested whether negative values are supported. They are in the data format, but they don't get visualised well. I do feel like we should support them. So lets go with `std::int64_t` as value type, realising this, I'll do a cleanup some day to convert all `unsigned int` type definitions to their `cstdint` flavours. No need to do that now.

The main question here is how do we want to represent both the counter name, and the series name. We could put both in the `StaticStringTracker`. That way the signature becomes something like:
```cpp
void counter_event(const unsigned int counter_id, const unsigned int series_id, const std::uint64_t counter_value);
```
Downside here is that ideally the name strings should be compile time constant to leverage the compile time constant crc. This would be the fastest way to deal with these counters, but it makes them less flexible and doesn't really allow us to add or remove series at runtime.

File: scalopus_tracing/src/native/native_tracepoint.cpp [hunk 2]
```diff
@@ -30,6 +30,8 @@
 #include <time.h>
 #include <iostream>
 
+#include <nlohmann/json.hpp>
```
Comment (iwanders) [inline scalopus_tracing/src/native/native_tracepoint.cpp]: Hmm, I'm surprised this worked without changes to the `CMakeLists.txt`. I thought I removed [this](https://github.com/iwanders/scalopus/blob/master/scalopus_tracing/CMakeLists.txt#L109) entry to `nlohmann::json`. I don't think it's used anywhere anymore...

File: scalopus_tracing/src/native/tracepoint_collector_native.h [hunk 3]
```diff
@@ -42,15 +44,17 @@ namespace tracepoint_collector_types
 //! Timepoint of that clock.
 using TimePoint = uint64_t;
 //! Trace event as it is stored in the ringbuffer.
-using StaticTraceEvent = std::tuple<TimePoint, unsigned int, uint8_t>;
+using StaticTraceEvent = std::tuple<TimePoint, unsigned int, uint8_t, nlohmann::json>;
```
Comment (iwanders) [inline scalopus_tracing/src/native/tracepoint_collector_native.h]: We should use a `std::unique_ptr<scalopus::Data>` as dynamic data payload. That allows preventing the necessity to require the `nlohmann::json` dependency in the process under test. Using a `unique_ptr` is the lightest thing I can think of. It default initialises to a `nullptr`, which is no dynamic data associated to the event. Or it can point to a payload which we can directly convert into cbor for the consumer side to interpret in whatever way it wants.

Not caused by this PR, but reading this back now it would've been much better if I had used some `using` things for the `unsigned int` and `uint8_t` here.

File: scalopus_tracing/include/scalopus_tracing/internal/counter_tracepoint.h (hunk 1 above)
Comment (iwanders) [reply scalopus_tracing/include/scalopus_tracing/internal/counter_tracepoint.h]: `void counter_event(const unsigned int counter_id, const std::int64_t counter_value);`

Comment (efernandez) [reply scalopus_tracing/include/scalopus_tracing/internal/counter_tracepoint.h]: Where `counter_id` would map to a string separated by `/`

File: scalopus_tracing/src/native/native_tracepoint.cpp (hunk 2 above)
Comment (iwanders) [reply scalopus_tracing/src/native/native_tracepoint.cpp]: Removed this dependency from this library.

**Comment by iwanders** [comment]
@efernandez , @jasonimercer

This should be ready for review now. All feedback has been incorporated and documentation has been updated.

File: scalopus_examples/src/showcase_count_events.cpp [hunk 4]
```diff
@@ -27,62 +27,48 @@
   OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
-#include "scalopus_tracing/tracing.h"
```
Comment (efernandez) [inline scalopus_examples/src/showcase_count_events.cpp]: Has this file actually been renamed? I wonder why.

File: scalopus_examples/src/showcase_count_events.cpp [hunk 5]
```diff
@@ -27,62 +27,48 @@
...
-    TRACE_MARK_EVENT_GLOBAL("foo");
-    TRACE_MARK_EVENT_GLOBAL("foo");
-    TRACE_MARK_EVENT_PROCESS("bar");
-    TRACE_MARK_EVENT_PROCESS("bar");
-    TRACE_MARK_EVENT_THREAD("buz");
-    TRACE_MARK_EVENT_THREAD("buz");
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    steps++;
+    t += 0.1;
+    std::int64_t a = static_cast<std::int64_t>(100.0 * std::sin(t * 10.0) + 100.0);
```
Comment (efernandez) [inline scalopus_examples/src/showcase_count_events.cpp]: pedantic: rename `a` and `b` to `slow` and `fast` and use the same order before in the `TRACE_COUNT_SERIES`

File: scalopus_tracing/include/scalopus_tracing/trace_macro.h [hunk 6]
```diff
@@ -66,4 +66,9 @@
 #define TRACE_MARK_EVENT_PROCESS(name) TRACE_MARK_EVENT_NAMED_ID(PROCESS, name, SCALOPUS_TRACKED_TRACE_ID_STRING(name))
 #define TRACE_MARK_EVENT_THREAD(name) TRACE_MARK_EVENT_NAMED_ID(THREAD, name, SCALOPUS_TRACKED_TRACE_ID_STRING(name))
 
+// Macro to set a counter series value.
+#define TRACE_COUNT_SERIES(name, series_name, value) TRACE_COUNT_SERIES_EVENT_NAMED(name "/" series_name, value)
+
+// Macro to set a counter value, just one series called 'count'.
+#define TRACE_COUNT(name, value) TRACE_COUNT_SERIES(name, "", value)
```
Comment (efernandez) [inline scalopus_tracing/include/scalopus_tracing/trace_macro.h]: If I understand correctly, you'll get `name/` in this case. Is that what we'll see in the .json file and the viewer?

File: scalopus_tracing/src/lttng/lttng_source.cpp [hunk 7]
```diff
@@ -131,10 +137,41 @@ std::vector<json> LttngSource::convertEvents()
         entry["ph"] = "i";
         entry["s"] = "t";
       }
+      else if (event.name() == "counter_event")
```
Comment (efernandez) [inline scalopus_tracing/src/lttng/lttng_source.cpp]: It seems that in the rest of the code you call them `count events` but here it's called `counter_event`. Maybe change the event name :thinking:

File: scalopus_tracing/src/lttng/lttng_source.cpp [hunk 8]
```diff
@@ -131,10 +137,41 @@ std::vector<json> LttngSource::convertEvents()
...
+      {
+        entry["ph"] = "C";
+        auto counter_series = LttngProvider::splitCounterSeriesName(trace_id_string);
+        entry["name"] = counter_series.first;
+        std::int64_t z = static_cast<std::int64_t>(event.eventData().at("value"));
```
Comment (efernandez) [inline scalopus_tracing/src/lttng/lttng_source.cpp]: `const auto z = static_cast<std::int64_t>(event.eventData().at("value"));

File: scalopus_tracing/src/lttng/lttng_source.cpp [hunk 9]
```diff
@@ -131,10 +137,41 @@ std::vector<json> LttngSource::convertEvents()
         entry["ph"] = "i";
         entry["s"] = "t";
       }
+      else if (event.name() == "counter_event")
+      {
+        entry["ph"] = "C";
+        auto counter_series = LttngProvider::splitCounterSeriesName(trace_id_string);
```
Comment (efernandez) [inline scalopus_tracing/src/lttng/lttng_source.cpp]: `const auto counter_series = LttngProvider::splitCounterSeriesName(trace_id_string);`

File: scalopus_tracing/src/lttng/lttng_source.cpp [hunk 10]
```diff
@@ -131,10 +137,41 @@ std::vector<json> LttngSource::convertEvents()
...
   }
 
+  // Sort res by "ts" because sometimes the "E"nd events are out of order wrt the "B"egin, leading to unfinished scope
+  // events.
+  std::stable_sort(result.begin(), result.end(), [](const json& lhs, const json& rhs) {
+    return lhs.at("ts").get<double>() < rhs.at("ts").get<double>();
+  });
+
+  // Need a reverse iteration here, to populate all counters with all series seen in the entire interval.
+  CounterMap counte_all_series;
```
Comment (efernandez) [inline scalopus_tracing/src/lttng/lttng_source.cpp]: `counte`?!

File: scalopus_tracing/src/lttng/lttng_source.cpp [hunk 11]
```diff
@@ -131,10 +137,41 @@ std::vector<json> LttngSource::convertEvents()
...
+  for (auto it = result.rbegin(); it < result.rend(); it++)
```
Comment (efernandez) [inline scalopus_tracing/src/lttng/lttng_source.cpp]: Why not `for (const auto& entry : result)`?

File: scalopus_tracing/src/native/native_trace_source.cpp [hunk 12]
```diff
@@ -138,6 +145,17 @@ std::vector<json> NativeTraceSource::finishInterval()
           entry["ph"] = "i";
           entry["s"] = "t";
         }
+        else if (type == TracePointCollectorNative::COUNTER)
+        {
+          entry["ph"] = "C";
+          auto counter_series = NativeTraceProvider::splitCounterSeriesName(trace_id_string);
```
Comment (efernandez) [inline scalopus_tracing/src/native/native_trace_source.cpp]: `const auto counter_series = NativeTraceProvider::splitCounterSeriesName(trace_id_string);`

File: scalopus_tracing/src/native/native_trace_source.cpp [hunk 13]
```diff
@@ -149,6 +167,23 @@ std::vector<json> NativeTraceSource::finishInterval()
     return lhs.at("ts").get<double>() < rhs.at("ts").get<double>();
   });
 
+  // Need a reverse iteration here, to populate all counters with all series seen in the entire interval.
+  ProcessCounter counte_all_series;
+  for (auto it = res.rbegin(); it < res.rend(); it++)
```
Comment (efernandez) [inline scalopus_tracing/src/native/native_trace_source.cpp]: Same comment as for lttng implementation

File: scalopus_tracing/src/scope_tracing_provider.cpp [hunk 14]
```diff
@@ -96,4 +96,21 @@ std::string ScopeTracingProvider::getScopeName(const ProcessTraceMap& mapping, c
...
+  auto index = trace_string.find("/");
+  if ((index != std::string::npos) && (index + 1 <= trace_string.size()))
+  {
+    res.first = trace_string.substr(0, index);
+    res.second = std::string{ std::next(trace_string.begin(), static_cast<long>(index + 1)), trace_string.end() };
+  }
+  else
+  {
+    res.first = trace_string;
+    res.second = "count";
```
Comment (efernandez) [inline scalopus_tracing/src/scope_tracing_provider.cpp]: I thought this would have been `res.second = "";` :thinking:

File: scalopus_tracing/src/spsc_ringbuffer.h [hunk 15]
```diff
@@ -90,6 +90,7 @@ class SPSCRingBuffer
     }
 
     container_[write_index] = std::move(v);  // move it into the container.
+    v.~ValueType();
```
Comment (efernandez) [inline scalopus_tracing/src/spsc_ringbuffer.h]: What is this explicit call for the destructor for?

Comment (efernandez) [reply scalopus_tracing/src/spsc_ringbuffer.h]: Your two last commits went in before I sent my review. nvm

File: scalopus_tracing/include/scalopus_tracing/trace_macro.h (hunk 6 above)
Comment (efernandez) [reply scalopus_tracing/include/scalopus_tracing/trace_macro.h]: Your two last commits went in before I sent my review. nvm

File: scalopus_examples/src/showcase_count_events.cpp (hunk 4 above)
Comment (iwanders) [reply scalopus_examples/src/showcase_count_events.cpp]: Lol... thanks Github...

So I removed `scalopus_tracing/test/test_native_tracing_macros.cpp` and `scalopus_tracing/test/test_lttng_tracing_macros.cpp`. I still have that test, but the file is called `scalopus_tracing/test/test_tracing_macros.cpp` and just compiled twice linking against different libraries.

I also introduced this `showcase_count_events.cpp`, which apparently looks enough like the trace macro test to be identified as a move? :o And then the headers got sorted. And I changed the `""` to `<>` include.

File: scalopus_tracing/src/lttng/lttng_source.cpp (hunk 7 above)
Comment (iwanders) [reply scalopus_tracing/src/lttng/lttng_source.cpp]: The event name is actually already a `count_event`, just forgot to change the comparison here. Good catch.

File: scalopus_examples/src/showcase_count_events.cpp (hunk 5 above)
Comment (iwanders) [reply scalopus_examples/src/showcase_count_events.cpp]: Yes, that's better.

File: scalopus_tracing/src/lttng/lttng_source.cpp (hunk 11 above)
Comment (iwanders) [reply scalopus_tracing/src/lttng/lttng_source.cpp]: It's a reverse iterator. `result.rbegin()`, so can't do an auto for loop.

File: scalopus_tracing/src/scope_tracing_provider.cpp (hunk 14 above)
Comment (iwanders) [reply scalopus_tracing/src/scope_tracing_provider.cpp]: I'm not sure if the UI supports empty series... I feel like it's better to have an explicit field name here?

File: scalopus_tracing/src/native/native_trace_source.cpp (hunk 13 above)
Comment (iwanders) [reply scalopus_tracing/src/native/native_trace_source.cpp]: Same reasoning, it's a reverse iterator.

File: scalopus_tracing/src/lttng/lttng_source.cpp (hunk 10 above)
Comment (iwanders) [reply scalopus_tracing/src/lttng/lttng_source.cpp]: Hah, typo and then copy pasted that without looking at it :)

File: scalopus_tracing/src/lttng/lttng_source.cpp (hunk 11 above)
Comment (efernandez) [reply scalopus_tracing/src/lttng/lttng_source.cpp]: Oh, that tiny `r`! :wink: 

:+1:

File: scalopus_tracing/src/scope_tracing_provider.cpp (hunk 14 above)
Comment (efernandez) [reply scalopus_tracing/src/scope_tracing_provider.cpp]: Oh, you're right. nvm

**Comment by jasonimercer** [comment]
:+1:
